utility/unit_of_work/uow_leader.go

```diff
@@ -0,0 +1,102 @@
+package unit_of_work
+
+import (
+	"github.com/pokt-network/pocket/logger"
+	coreTypes "github.com/pokt-network/pocket/shared/core/types"
+	"github.com/pokt-network/pocket/shared/modules"
+)
+
+var (
+	_ modules.UtilityUnitOfWork       = &leaderUtilityUnitOfWork{}
+	_ modules.LeaderUtilityUnitOfWork = &leaderUtilityUnitOfWork{}
+)
+
+type leaderUtilityUnitOfWork struct {
+	baseUtilityUnitOfWork
+}
+
+func NewForLeader(height int64, readContext modules.PersistenceReadContext, rwPersistenceContext modules.PersistenceRWContext) *leaderUtilityUnitOfWork {
+	return &leaderUtilityUnitOfWork{
+		baseUtilityUnitOfWork: baseUtilityUnitOfWork{
+			height:                 height,
+			persistenceReadContext: readContext,
+			persistenceRWContext:   rwPersistenceContext,
+			logger:                 logger.Global.CreateLoggerForModule(leaderUtilityUOWModuleName),
+		},
+	}
+}
+
+func (uow *leaderUtilityUnitOfWork) CreateAndApplyProposalBlock(proposer []byte, maxTxBytes uint64) (stateHash string, txs [][]byte, err error) {
+	prevBlockByzantineVals, err := uow.prevBlockByzantineValidators()
+	if err != nil {
+		return "", nil, err
+	}
+
+	// begin block lifecycle phase
+	if err := uow.beginBlock(prevBlockByzantineVals); err != nil {
+		return "", nil, err
+	}
+	txs = make([][]byte, 0)
+	txsTotalBz := uint64(0)
+	txIdx := 0
+
+	mempool := uow.GetBus().GetUtilityModule().GetMempool()
+	for !mempool.IsEmpty() {
+		// NB: In order for transactions to have entered the mempool, `HandleTransaction` must have
+		// been called which handles basic checks & validation.
+		txBz, err := mempool.PopTx()
+		if err != nil {
+			return "", nil, err
+		}
+
+		tx, err := coreTypes.TxFromBytes(txBz)
+		if err != nil {
+			return "", nil, err
+		}
+
+		txBzSize := uint64(len(txBz))
+		txsTotalBz += txBzSize
+
+		// Exceeding maximum transaction bytes to be added in this block
+		if txsTotalBz >= maxTxBytes {
+			// Add back popped tx to be applied in a future block
+			if err := mempool.AddTx(txBz); err != nil {
+				return "", nil, err
+			}
+			break // we've reached our max
+		}
+
+		txResult, err := uow.hydrateTxResult(tx, txIdx)
+		if err != nil {
+			uow.logger.Err(err).Msg("Error in ApplyTransaction")
+			// TODO(#327): Properly implement 'unhappy path' for save points
+			if err := uow.revertLastSavePoint(); err != nil {
+				return "", nil, err
+			}
+			txsTotalBz -= txBzSize
+			continue
+		}
+
+		// Index the transaction
+		if err := uow.persistenceRWContext.IndexTransaction(txResult); err != nil {
+			uow.logger.Fatal().Err(err).Msgf("TODO(#327): The transaction can by hydrated but not indexed. Crash the process for now: %v\n", err)
+		}
+
+		txs = append(txs, txBz)
+		txIdx++
+	}
+
+	if err := uow.endBlock(proposer); err != nil {
+		return "", nil, err
+	}
+
+	// TODO: @deblasis - this should be from a ReadContext (the ephemeral/staging one)
```

Comment:
I disagree (or am missing something).

1. The `persistenceRWContext` is the one that has all the changes in the block being built, applied and proposed by the leader.
2. The `stateHash` computed here happens in the beginning of the consensus lifecycle that's distributed throughout the network as a proposal.
3. Since the changes in `persistenceRWContext` are not committed, the `readContext` (a separate context) would have a different state hash.

What am I missing?

Reply:
> What am I missing?
I guess this: https://github.com/pokt-network/pocket/pull/577#issuecomment-1472747893